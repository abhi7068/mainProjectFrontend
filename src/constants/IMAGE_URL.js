const IMAGE_URL = 'https://images.freeimages.com/images/small-previews/f9d/cup-and-saucer-1639822.jpg'
export default IMAGE_URL;
