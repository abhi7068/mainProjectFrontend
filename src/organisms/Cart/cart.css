a {
  text-decoration: none;
  color: black;
}

.remove-button {
  color: grey;
}
.remove-button:hover {
  color: red;
}
.main {
  margin-bottom: 2em;
}
.root {
  display: flex;
  flex-direction: row;
  flex-grow: 1;
  margin-bottom: 1em;
  align-items: center;
}
.cardRoot {
  flex-grow: 1;
  display: flex;
  flex-direction: column;
}
.grid {
  margin: 3em 1em;
  flex-grow: 1;
}
.cardBody {
  flex-grow: 1;
  display: flex;
  justify-content: space-around;
}
.details {
  display: flex;
  justify-content: center;
  flex-direction: column;
}
.coverDiv {
  max-width: 250px;
  width: 250px;
  min-width: 250px;
  display: flex;
  align-items: center;
}
.cover {
  width: 100%;
}
.controls {
  display: flex;
  justify-content: center;
  align-items: center;
}
.playIcon {
  height: 38;
  width: 38;
}
@media screen and (max-width: 600px) {
  .root {
    flex-direction: column !important;
  }
  .cardBody {
    flex-direction: column !important;
  }
}
