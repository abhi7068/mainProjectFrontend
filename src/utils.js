import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import { createMemoryHistory } from 'history';
import thunk from 'redux-thunk';
import PropTypes from 'prop-types';
import reducer from '../src/store/reducers/rootReducer';
import { Router } from 'react-router-dom';

function render(
  ui,
  {
    initialState,
    store = createStore(reducer, initialState, applyMiddleware(thunk)),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    const history = createMemoryHistory();
    return (
      <Provider store={store}>
        <Router history={history}>{children}</Router>
      </Provider>
    );
  }
  Wrapper.propTypes = {
    children: PropTypes.elementType,
  };
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}
export * from '@testing-library/react';
export { render };
