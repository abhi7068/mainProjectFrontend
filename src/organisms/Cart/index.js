import React from 'react';
// import _ from 'lodash';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import AddIcon from '@material-ui/icons/Add';
import RemoveIcon from '@material-ui/icons/Remove';
import Grid from '@material-ui/core/Grid';
import CardActions from '@material-ui/core/CardActions';
import Button from '@material-ui/core/Button';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Chip from '@material-ui/core/Chip';
import Alert from '@material-ui/lab/Alert';
import { getUserById } from '../../services/UserService';
import CancelOutlinedIcon from '@material-ui/icons/CancelOutlined';
import {
  removeOne,
  addOne,
  removeProduct,
} from '../../store/actions/cartAction';
import { productDetails } from '../../store/actions/productDetailAction';
import { Link, withRouter } from 'react-router-dom';
import IMAGE_URL from '../../constants/IMAGE_URL';
import kebab from '../../services/kebab';
import { deleteOneApi, addOneApi } from '../../services/cartServices';
import { getUserId } from '../../constants/BASE_URL';
import './cart.css';
import SnackBar from '../../atoms/SnackBar';

export class Cart extends React.Component {
  constructor(props) {
    super(props);
    this.state = { error: '', open: false };
    this.proceedToGift = this.proceedToGift.bind(this);
  }

  componentDidMount() {
    const details = JSON.parse(localStorage.getItem('storeResponseDetails'));
    if (details) {
      this.props.getUserById(details.id);
    }

    // getUserById(details.id);
  }

  proceedToGift() {
    const details = JSON.parse(localStorage.getItem('storeResponseDetails'));
    const totalPrice = this.props.cart.reduce(
      (total, item) => total + item.product.price * item.count,
      0
    );
    const {
      singleUserDetail: {
        data: { yoyobalance },
      },
    } = this.props;
    if (details) {
      if (totalPrice <= yoyobalance) {
        const balanceLeft = yoyobalance - totalPrice;
        this.props.history.push('/send/gift/' + balanceLeft);
      } else {
        this.setState({ open: true });
        setTimeout(() => {
          this.setState({
            open: false,
          });
        }, 1000);
      }
    } else {
      this.props.history.push('/login');
    }
  }

  render() {
    return (
      <div className="grid" data-testid="cart-container">
        {this.state.open && (
          <SnackBar
            path="/"
            name="you don't have sufficient yoyo points"
            severity="warning"
          />
        )}
        {this.state.error ? (
          <Alert severity="warning">{this.state.error}</Alert>
        ) : (
          ''
        )}
        {this.props.cart && this.props.cart.length ? (
          <Card className="main">
            <Typography
              variant="h4"
              component="h2"
              style={{ textAlign: 'left', padding: '1em 1em 0em 1em' }}
            >
              Your Cart
            </Typography>
            <Grid
              container
              spacing={3}
              style={{ width: '100%', margin: '0px' }}
            >
              <Grid item xs={12} sm={12} md={8}>
                {this.props.cart.map((item, index) => (
                  <Card
                    key={item.product._id}
                    className="root"
                    variant="outlined"
                  >
                    <div className="coverDiv">
                      <img
                        data-testid="product-img"
                        src={IMAGE_URL}
                        className="cover"
                        alt="Product Title"
                      />
                    </div>
                    <div className="cardBody">
                      <div className="details">
                        <CardContent>
                          <Typography
                            data-testid="product-heading"
                            component="h5"
                            variant="h5"
                            onClick={() => {
                              this.props.productDetails(item.product._id);
                            }}
                          >
                            <Link
                              to={`/${kebab(
                                item.product.category &&
                                  item.product.category.title
                                  ? item.product.category.title
                                  : 'No Category'
                              )}/${kebab(item.product.product_name)}`}
                            >
                              {item.product.product_name}
                            </Link>
                          </Typography>
                          <Typography variant="subtitle1" color="textSecondary">
                            <Chip
                              data-testId="category-name"
                              variant="outlined"
                              color="secondary"
                              size="small"
                              label={
                                item.product.category &&
                                item.product.category.title
                                  ? item.product.category.title
                                  : 'No Category'
                              }
                            />
                          </Typography>
                        </CardContent>
                        <div className="controls" data-testid="counter">
                          <IconButton
                            data-testid="decrease"
                            disabled={item.count <= 1}
                            onClick={() => {
                              if (getUserId()) {
                                deleteOneApi(item._id)
                                  .then(() => {
                                    this.setState({ error: '' });
                                  })
                                  .catch(() => {
                                    this.setState({
                                      error: 'Could Not Decrease Quantity',
                                    });
                                    setTimeout(() => {
                                      this.setState({
                                        error: '',
                                      });
                                    }, 5000);

                                    this.props.addOne(index);
                                  });
                              }
                              this.props.removeOne(index);
                            }}
                          >
                            <RemoveIcon />
                          </IconButton>
                          {item.count}
                          <IconButton
                            data-testid="increase"
                            onClick={() => {
                              if (getUserId()) {
                                addOneApi(item._id)
                                  .then(() => {
                                    this.setState({ error: '' });
                                  })
                                  .catch(() => {
                                    this.setState({
                                      error: 'Could Not Increase Quantity',
                                    });
                                    setTimeout(() => {
                                      this.setState({
                                        error: '',
                                      });
                                    }, 5000);
                                    this.props.removeOne(index);
                                  });
                              }
                              this.props.addOne(index);
                            }}
                          >
                            <AddIcon />
                          </IconButton>
                        </div>
                      </div>
                      <div
                        style={{
                          display: 'flex',
                          flexDirection: 'column',
                          justifyContent: 'center',
                        }}
                      >
                        <div>
                          {item.product.discount ? (
                            <>
                              <Typography variant="h6">
                                Price:{' '}
                                <span
                                  style={{
                                    color: '#B12704',
                                    textDecoration: 'line-through',
                                  }}
                                >
                                  {item.product.price} points
                                </span>
                              </Typography>
                              <Typography variant="h6">
                                Discount:{' '}
                                <span style={{ color: '#B12704' }}>
                                  {item.product.discount}%
                                </span>
                              </Typography>

                              <Typography variant="h6">
                                New Price:{' '}
                                <strong style={{ color: 'green' }}>
                                  {(item.product.price *
                                    (100 - item.product.discount)) /
                                    100}{' '}
                                  points
                                </strong>
                              </Typography>
                            </>
                          ) : (
                            <Typography variant="h6">
                              Price:{' '}
                              <strong
                                style={{
                                  color: 'green',
                                }}
                              >
                                {item.product.price} points
                              </strong>
                            </Typography>
                          )}
                        </div>
                      </div>
                    </div>
                    <div
                      style={{
                        display: 'flex',
                        flexDirection: 'column',
                        justifyContent: 'center',
                      }}
                    >
                      <IconButton
                        data-testid="remove-button"
                        className="remove-button"
                        onClick={() => {
                          this.props.removeProduct(item.product._id);
                        }}
                      >
                        <CancelOutlinedIcon />
                      </IconButton>
                    </div>
                  </Card>
                ))}
                <Card variant="outlined">
                  <Typography
                    data-testid="subtotal2"
                    variant="h5"
                    component="h2"
                    style={{ textAlign: 'right', paddingRight: '1em' }}
                  >
                    {' '}
                    Subtotal (
                    {this.props.cart.reduce(
                      (total, item) => total + item.count,
                      0
                    )}{' '}
                    items) :{' '}
                    <span style={{ color: 'green' }}>
                      {this.props.cart.reduce(
                        (total, item) =>
                          total +
                          Math.ceil(
                            (item.product.price *
                              item.count *
                              (100 - item.product.discount)) /
                              100
                          ),
                        0
                      )}{' '}
                      points
                    </span>
                  </Typography>
                </Card>
              </Grid>
              <Grid item xs={12} sm={12} md={4}>
                <Card variant="outlined" className="cardRoot">
                  <CardContent>
                    <Typography
                      variant="h5"
                      component="h2"
                      className="title"
                      data-testid="subtotal1"
                    >
                      Subtotal (
                      {this.props.cart.reduce(
                        (total, item) => total + item.count,
                        0
                      )}{' '}
                      items) :{' '}
                      <span
                        style={{
                          color: 'green',
                        }}
                      >
                        {this.props.cart.reduce(
                          (total, item) =>
                            total +
                            Math.ceil(
                              (item.product.price *
                                item.count *
                                (100 - item.product.discount)) /
                                100
                            ),
                          0
                        )}{' '}
                        points
                      </span>
                    </Typography>
                    <Typography>This order contains a gift</Typography>
                  </CardContent>

                  <CardActions style={{ justifyContent: 'center' }}>
                    <Button
                      size="small"
                      variant="contained"
                      color="primary"
                      onClick={this.proceedToGift}
                    >
                      Proceed To Gift
                    </Button>
                  </CardActions>
                </Card>
              </Grid>
            </Grid>
          </Card>
        ) : (
          <Typography variant="h5" component="h6">
            Please Add Something To Cart
          </Typography>
        )}
      </div>
    );
  }
}

Cart.propTypes = {
  cart: PropTypes.array,
  userDetails: PropTypes.object,
  addOne: PropTypes.func,
  removeOne: PropTypes.func,
  productDetails: PropTypes.func,
  removeProduct: PropTypes.func,
  getUserById: PropTypes.func,
  singleUserDetail: PropTypes.string,
  history: PropTypes.object,
};
/* function mapStateToProps(state) {

  return {
    categories: state.categories,
  };
} */

export default withRouter(
  connect(
    ({ cart, userDetails, singleUserDetail }) => ({
      cart,
      userDetails,
      singleUserDetail,
    }),
    {
      removeOne,
      productDetails,
      addOne,
      removeProduct,
      getUserById,
    }
  )(Cart)
);
