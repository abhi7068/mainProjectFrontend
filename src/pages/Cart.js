import React from 'react';
import Cart from '../organisms/Cart';
const CartPage = () => {
  return <Cart />;
};

export default CartPage;
