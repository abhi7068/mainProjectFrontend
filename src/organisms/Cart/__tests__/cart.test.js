import React from 'react';
import { render, cleanup, fireEvent } from '../../../utils';
import Cart from '../index';
import IMAGE_URL from '../../../constants/IMAGE_URL';

const cartNoCount = [
  {
    count: 1,
    _id: '5ea522c906d5053ee84d185f',
    product: {
      disabled: false,
      reviews: [],
      _id: '5e8c615afa2d330554c806c4',
      product_name: 'watch',
      description: 'great watch',
      price: 100,
      discount: 0,
      created_at: '2020-04-07T11:17:46.960Z',
      updated_at: '2020-04-07T11:17:46.960Z',
      __v: 0,
    },
  },
];

afterEach(cleanup);

test('should take a snapshot', () => {
  const { asFragment } = render(<Cart />, {
    initialState: { cart: cartNoCount },
  });
  expect(
    asFragment(render(<Cart />, { initialState: { cart: cartNoCount } }))
  ).toMatchSnapshot();
});

test('should take a snapshot', () => {
  const { asFragment } = render(<Cart />);
  expect(asFragment(render(<Cart />))).toMatchSnapshot();
});

test('should equal to 1', () => {
  const { getByTestId, getByText } = render(<Cart />, {
    initialState: { cart: cartNoCount },
  });
  expect(getByTestId('counter')).toHaveTextContent(1);
  expect(getByText('Your Cart')).toBeInTheDocument();
  expect(getByTestId('increase')).not.toHaveAttribute('disabled');
});

test('remove button should be disabled', () => {
  const { getByTestId } = render(<Cart />, {
    initialState: { cart: cartNoCount },
  });
  expect(getByTestId('counter')).toHaveTextContent(1);
  expect(getByTestId('decrease')).toBeDisabled();
});

test('should have product heading and image', () => {
  const { getByTestId } = render(<Cart />, {
    initialState: { cart: cartNoCount },
  });
  expect(getByTestId('product-img').src).toEqual(IMAGE_URL);
  expect(getByTestId('product-heading')).toHaveTextContent('watch');
});

test('should have category Name', () => {
  const { getByTestId } = render(<Cart />, {
    initialState: { cart: cartNoCount },
  });
  expect(getByTestId('category-name')).toHaveTextContent('No Category');
});

test('should increment to 2 and subtotal increases', () => {
  const { getByTestId } = render(<Cart />, {
    initialState: { cart: cartNoCount },
  });
  fireEvent.click(getByTestId('increase'));
  expect(getByTestId('counter')).toHaveTextContent(2);
  expect(getByTestId('subtotal1')).toHaveTextContent(
    'Subtotal (2 items) : 200 points'
  );
  expect(getByTestId('subtotal2')).toHaveTextContent(
    'Subtotal (2 items) : 200 points'
  );
});

test('should remove from cart', () => {
  const { getByTestId } = render(<Cart />, {
    initialState: { cart: cartNoCount },
  });
  fireEvent.click(getByTestId('remove-button'));
  expect(getByTestId('cart-container')).toHaveTextContent(
    'Please Add Something To Cart'
  );
});
