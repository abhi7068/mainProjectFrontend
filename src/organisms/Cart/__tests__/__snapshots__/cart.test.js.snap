// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid"
    data-testid="cart-container"
  >
    <div
      class="MuiPaper-root MuiCard-root main MuiPaper-elevation1 MuiPaper-rounded"
    >
      <h2
        class="MuiTypography-root MuiTypography-h4"
        style="text-align: left; padding: 1em 1em 0em 1em;"
      >
        Your Cart
      </h2>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
        style="width: 100%; margin: 0px;"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-8"
        >
          <div
            class="MuiPaper-root MuiCard-root root MuiPaper-outlined MuiPaper-rounded"
          >
            <div
              class="coverDiv"
            >
              <img
                alt="Product Title"
                class="cover"
                data-testid="product-img"
                src="https://images.freeimages.com/images/small-previews/f9d/cup-and-saucer-1639822.jpg"
              />
            </div>
            <div
              class="cardBody"
            >
              <div
                class="details"
              >
                <div
                  class="MuiCardContent-root"
                >
                  <h5
                    class="MuiTypography-root MuiTypography-h5"
                    data-testid="product-heading"
                  >
                    <a
                      href="/no-category/watch"
                    >
                      watch
                    </a>
                  </h5>
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
                  >
                    <div
                      class="MuiChip-root MuiChip-colorSecondary MuiChip-outlined MuiChip-outlinedSecondary MuiChip-sizeSmall"
                      data-testid="category-name"
                    >
                      <span
                        class="MuiChip-label MuiChip-labelSmall"
                      >
                        No Category
                      </span>
                    </div>
                  </h6>
                </div>
                <div
                  class="controls"
                  data-testid="counter"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                    data-testid="decrease"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 13H5v-2h14v2z"
                        />
                      </svg>
                    </span>
                  </button>
                  1
                  <button
                    class="MuiButtonBase-root MuiIconButton-root"
                    data-testid="increase"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
              <div
                style="display: flex; flex-direction: column; justify-content: center;"
              >
                <div>
                  <h6
                    class="MuiTypography-root MuiTypography-h6"
                  >
                    Price: 
                    <strong
                      style="color: green;"
                    >
                      100 points
                    </strong>
                  </h6>
                </div>
              </div>
            </div>
            <div
              style="display: flex; flex-direction: column; justify-content: center;"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root remove-button"
                data-testid="remove-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm3.59-13L12 10.59 8.41 7 7 8.41 10.59 12 7 15.59 8.41 17 12 13.41 15.59 17 17 15.59 13.41 12 17 8.41z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <div
            class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
          >
            <h2
              class="MuiTypography-root MuiTypography-h5"
              data-testid="subtotal2"
              style="text-align: right; padding-right: 1em;"
            >
               Subtotal (1 items) : 
              <span
                style="color: green;"
              >
                100 points
              </span>
            </h2>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-4"
        >
          <div
            class="MuiPaper-root MuiCard-root cardRoot MuiPaper-outlined MuiPaper-rounded"
          >
            <div
              class="MuiCardContent-root"
            >
              <h2
                class="MuiTypography-root title MuiTypography-h5"
                data-testid="subtotal1"
              >
                Subtotal (1 items) : 
                <span
                  style="color: green;"
                >
                  100 points
                </span>
              </h2>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                This order contains a gift
              </p>
            </div>
            <div
              class="MuiCardActions-root MuiCardActions-spacing"
              style="justify-content: center;"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Proceed To Gift
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should take a snapshot 2`] = `
<DocumentFragment>
  <div
    class="grid"
    data-testid="cart-container"
  >
    <h6
      class="MuiTypography-root MuiTypography-h5"
    >
      Please Add Something To Cart
    </h6>
  </div>
</DocumentFragment>
`;
